File: check_hosts.pl
#!/usr/bin/perl

use strict;
use warnings;

use Term::ANSIColor;
use Data::Dumper;

my $servers = `source ~/workshop-ansible/OpenStack/openrc.sh && nova list 2>&1 | grep -Ev "server00|server-00" | awk '{ print \$4 }' | grep server | tr '\n' ' '`;
my @servers = split / /, $servers;

foreach my $server (@servers)
{   
    my $rawload = `/root/DBaaS-TS/get-load-dbaas-ts.sh $server | awk '{ print \$2 }'`;

    if ( $rawload !~ /^\s*$/ )
    {
        my $load = sprintf("%.2f",$rawload);
        chomp ($load);
        
        my $running;
        if (grep /$server-fresh/, @servers) { $running = 1 } else { $running = 0 }

        if ( $load >= 0.75 )
        {   
            if ( $running == 0 )
            {   
                if ( $server =~ /fresh/ ) 
                {
                    printf("Host %-25s on fire (%s)  [but already 2nd gen, not spawn-looping]\n", colored($server, 'bold cyan'), colored($load, 'bold red'));
                }
                else
                {
                    printf("Host %-25s on fire (%s)  [spawning a fresh new instance]\n", colored($server, 'bold cyan'), colored($load, 'bold red'));
                    `source ~/workshop-ansible/OpenStack/openrc.sh && nova boot --flavor vps-ssd-1 --image 'Debian 8' --user-data ~/workshop-ansible/OpenStack/post-install.yaml $server-fresh 2>&1`;
                }
            }
            else
            {   
                printf("Host %-25s on fire (%s)  [but a clone was already spawned]\n", colored($server, 'bold cyan'), colored($load, 'bold red'));
            }
        }
        else 
        {   
            if ( $running == 0 )
            {   
                printf("Host %-25s is fine (%s)\n", colored($server, 'bold cyan'), colored($load, 'bold green'));
            }
            else
            {   
                printf("Host %-25s is fine (%s)  [so %s will be destroyed]\n", colored($server, 'bold cyan'), colored($load, 'bold green'), "$server-fresh");
                `source ~/workshop-ansible/OpenStack/openrc.sh && nova delete '$server-fresh' 2>&1`;
            }
        }
    } else { printf("Host %-25s not in DBaaS-TS [be patient, or do something!]\n", colored($server, 'bold cyan')) }
}

File: check_hosts.py
#!/usr/bin/env python

import sys
import re
import os
import commands
import base64

sys.path.insert(0, '/root/workshop-ansible/OpenStack/')
from credentials import get_nova_credentials_v2

sys.path.insert(0, '/root/workshop-ansible/SMS/')
from sms import send_sms

from termcolor import colored # python-termcolor
from novaclient.client import Client
from pprint import pprint

credentials = get_nova_credentials_v2()
nova_client = Client(**credentials)
servers     = nova_client.servers.list()
image       = nova_client.images.find(name="Debian 8")
flavor      = nova_client.flavors.find(name="vps-ssd-1")
data        = open("/root/workshop-ansible/OpenStack/post-install.yaml","r")

for server in servers:
    match = re.search(r"(\w*?-\d[1-9])", server.name)
    if match:
        rawload     = commands.getstatusoutput("/root/workshop-ansible/DBaaS-TS/get-load-dbaas-ts.sh " + server.name + " | awk '{ print $2 }'")
        server_cyan = colored(server.name, 'cyan' , attrs=['bold'])

        if rawload[1]:
            load        = float(rawload[1])
            load_red    = colored('{0:.2f}'.format(load), 'red'  , attrs=['bold'])
            load_green  = colored('{0:.2f}'.format(load), 'green', attrs=['bold'])
            fresh       = server.name + '-fresh'
            grep        = any(server.name == fresh for server in servers) 

            if load >= 0.75:
                if grep == False:
                    if re.search(r"fresh", server.name):
                        print("Host %-36s on fire (%s)  [but already 2nd gen, not spawn-looping]"% (server_cyan, load_red))
                    else:
                        print("Host %-36s on fire (%s)  [spawning a fresh new instance]"% (server_cyan, load_red))
                        server = nova_client.servers.create(name = server.name + '-fresh', 
                                                            image = image.id, 
                                                            flavor = flavor.id,
                                                            userdata = data)
                        sms = send_sms('Workload too high. Spawning new instance ' + fresh)
                else:
                    print("Host %-36s on fire (%s)  [but a clone was already spawned]"% (server_cyan, load_red))
            else:
                if grep == False:
                    print("Host %-36s is fine (%s)"% (server_cyan, load_green))
                else:
                    print("Host %-36s is fine (%s)  [so %s will be destroyed]"% (server_cyan, load_green, server.name + '-fresh'))
                    server = nova_client.servers.find(name=fresh)
                    nova_client.servers.delete(server.id)
                    sms = send_sms('Workload back to normal. Deleting ' + fresh)

        else:
            print("Host %-36s not in DBaaS-TS [be patient, or do something!]"% (server_cyan))
